File: src/main/java/com/somefrills/features/farming/autofarmer/CocoBeansMovement.java
package com.somefrills.features.farming.autofarmer;

import org.jspecify.annotations.NonNull;

/**
 * Coco beans pattern farming.
 * Cycles: Forward → Right → Backward → Right → Forward (loop)
 */
public class CocoBeansMovement implements MovementStrategy {
    private FarmingState currentState = FarmingState.FORWARD;

    public CocoBeansMovement() {
    }

    @NonNull
    @Override
    public MovementState getCurrentState() {
        return currentState.toMovementState();
    }

    @Override
    public void nextState() {
        currentState = (FarmingState) currentState.next();
    }

    private enum FarmingState implements MovementStrategy.State {
        FORWARD {
            @Override
            public FarmingState next() {
                return RIGHT;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(MovementState.FORWARD);
            }
        },
        RIGHT {
            @Override
            public FarmingState next() {
                return BACKWARD;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(MovementState.RIGHT);
            }
        },
        BACKWARD {
            @Override
            public FarmingState next() {
                return RIGHT_2;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(MovementState.BACKWARD);
            }
        },
        RIGHT_2 {
            @Override
            public FarmingState next() {
                return FORWARD;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(MovementState.RIGHT);
            }
        }
    }
}

File: src/main/kotlin/com/somefrills/features/farming/autofarmer/CocoBeansMovement.kt
package com.somefrills.features.farming.autofarmer

/**
 * Coco beans pattern farming.
 * Cycles: Forward → Right → Backward → Right → Forward (loop)
 */
class CocoBeansMovement : MovementStrategy() {
    private var state: FarmingState = FarmingState.FORWARD

    override fun getState(): MovementState {
        return state.toMovementState()
    }

    override fun nextState() {
        state = state.next() as FarmingState
    }

    private enum class FarmingState : State {
        FORWARD {
            override fun next() = RIGHT
            override fun toMovementState() = MovementState(MovementState.FORWARD)
        },
        RIGHT {
            override fun next() = BACKWARD
            override fun toMovementState() = MovementState(MovementState.RIGHT)
        },
        BACKWARD {
            override fun next() = RIGHT_2
            override fun toMovementState() = MovementState(MovementState.BACKWARD)
        },
        RIGHT_2 {
            override fun next() = FORWARD
            override fun toMovementState() = MovementState(MovementState.RIGHT)
        };
    }
}